File: apps/portal/src/lib/api/schemas/compliance.ts
import { z } from 'zod';

export const CheckResultItemSchema = z.object({
  rule_id: z.string(),
  article: z.string(),
  element_global_id: z.string(),
  element_type: z.union([z.string(), z.null()]).optional(),
  element_name: z.union([z.string(), z.null()]).optional(),
  status: z.enum(['pass', 'fail', 'warn', 'skip', 'error']),
  message: z.string(),
  actual_value: z.union([z.string(), z.number(), z.boolean(), z.null()]).optional(),
  expected_value: z.union([z.string(), z.number(), z.boolean(), z.null()]).optional(),
  property_set: z.union([z.string(), z.null()]).optional(),
  property_name: z.union([z.string(), z.null()]).optional(),
  severity: z.string(),
});

export type CheckResultItem = z.infer<typeof CheckResultItemSchema>;

export const RuleSummaryItemSchema = z.object({
  rule_id: z.string(),
  article: z.string(),
  titles: z.record(z.string(), z.string()).optional(),
  title: z.string().nullish(),
  title_nl: z.string().nullish(),
  category: z.string(),
  severity: z.string(),
  total_checked: z.number(),
  passed: z.number(),
  failed: z.number(),
  warned: z.number(),
  skipped: z.number(),
  errors: z.number(),
});

export type RuleSummaryItem = z.infer<typeof RuleSummaryItemSchema>;

export const CategorySummaryItemSchema = z.object({
  category: z.string(),
  total_rules: z.number(),
  total_checks: z.number(),
  passed: z.number(),
  failed: z.number(),
  warned: z.number(),
});

export type CategorySummaryItem = z.infer<typeof CategorySummaryItemSchema>;

export const ComplianceCheckResponseSchema = z.object({
  file_id: z.string(),
  job_id: z.string().uuid(),
  checked_at: z.string(),
  total_rules: z.number(),
  total_elements_checked: z.number(),
  rules_summary: z.array(RuleSummaryItemSchema),
  category_summary: z.array(CategorySummaryItemSchema),
  details: z.array(CheckResultItemSchema),
});

export type ComplianceCheckResponse = z.infer<typeof ComplianceCheckResponseSchema>;

export const ComplianceSummaryResponseSchema = z.object({
  file_id: z.string(),
  job_id: z.string().uuid(),
  checked_at: z.string(),
  total_rules: z.number(),
  total_elements_checked: z.number(),
  rules_summary: z.array(RuleSummaryItemSchema),
  category_summary: z.array(CategorySummaryItemSchema),
});

export type ComplianceSummaryResponse = z.infer<typeof ComplianceSummaryResponseSchema>;

export const ComplianceFrameworkEnum = z.enum(['bbl', 'wkb']);

export type ComplianceFramework = z.infer<typeof ComplianceFrameworkEnum>;

export const ProjectComplianceReportItemSchema = z.object({
  job_id: z.string().uuid(),
  file_id: z.string().uuid(),
  document_id: z.string().uuid(),
  document_name: z.string(),
  document_discipline: z.string(),
  file_name: z.string(),
  file_version: z.number().int(),
  framework: ComplianceFrameworkEnum,
  checked_at: z.string(),
  finished_at: z.string(),
  pass_count: z.number().int(),
  warn_count: z.number().int(),
  fail_count: z.number().int(),
  total_rules: z.number().int(),
  total_elements_checked: z.number().int(),
  overall_score: z.number().int(),
});

export type ProjectComplianceReportItem = z.infer<typeof ProjectComplianceReportItemSchema>;

export const ProjectComplianceReportListSchema = z.object({
  items: z.array(ProjectComplianceReportItemSchema),
});

export type ProjectComplianceReportList = z.infer<typeof ProjectComplianceReportListSchema>;

File: apps/api/src/bimdossier_api/schemas/compliance.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class ComplianceCheckRequest(BaseModel):
    # None = derive from the project's building_type (the default). Pass an
    # explicit value to override rule filtering for this one check.
    building_type: str | None = None
    categories: list[str] | None = None
    framework: str = "bbl"


class CheckResultItem(BaseModel):
    rule_id: str
    article: str
    element_global_id: str
    element_type: str | None = None
    element_name: str | None = None
    status: Literal["pass", "fail", "warn", "skip", "error"]
    message: str
    actual_value: str | float | int | bool | None = None
    expected_value: str | float | int | bool | None = None
    property_path: str | None = None
    severity: str


class RuleSummaryItem(BaseModel):
    rule_id: str
    article: str
    title: str
    title_nl: str
    category: str
    severity: str
    total_checked: int
    passed: int
    failed: int
    warned: int
    skipped: int
    errors: int


class CategorySummaryItem(BaseModel):
    category: str
    total_rules: int
    total_checks: int
    passed: int
    failed: int
    warned: int


class ArbiterComplianceResult(BaseModel):
    """Strict validation gate for the Arbiter's ``check_compliance`` payload.

    The structural keys must be PRESENT — ``routers/compliance.py`` reads the
    result with ``result.get(key, <empty default>)``, so a malformed/empty
    Arbiter payload would otherwise be persisted as a SUCCEEDED but 0-rule
    "clean" report (a silent false-pass). Requiring the keys here forces a
    ``ComplianceCheckError`` instead. The nested item types are the same models
    the response is built from, so a payload that validates here is guaranteed to
    build a ``ComplianceCheckResponse``. Extra keys the Arbiter sends (``file_id``,
    ``framework``, ``details[].reasoning``, ``format_coverage``) are tolerated;
    the caller stores the original dict, so nothing is dropped.
    """

    model_config = ConfigDict(extra="allow")

    checked_at: str
    total_rules: int
    total_elements_checked: int
    rules_summary: list[RuleSummaryItem]
    category_summary: list[CategorySummaryItem]
    details: list[CheckResultItem]


class ComplianceCheckResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    file_id: str
    job_id: UUID
    framework: str = "bbl"
    checked_at: str
    total_rules: int
    total_elements_checked: int
    rules_summary: list[RuleSummaryItem]
    category_summary: list[CategorySummaryItem]
    details: list[CheckResultItem]


class ComplianceSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    file_id: str
    job_id: UUID
    framework: str = "bbl"
    checked_at: str
    total_rules: int
    total_elements_checked: int
    rules_summary: list[RuleSummaryItem]
    category_summary: list[CategorySummaryItem]


class ProjectComplianceReportItem(BaseModel):
    """One row in the project-level compliance reports list — latest succeeded
    job per (file_id, framework)."""

    model_config = ConfigDict(from_attributes=True)

    job_id: UUID
    file_id: UUID
    document_id: UUID
    document_name: str
    document_discipline: str
    file_name: str
    file_version: int
    framework: str
    checked_at: str
    finished_at: datetime
    pass_count: int
    warn_count: int
    fail_count: int
    total_rules: int
    total_elements_checked: int
    overall_score: int  # 0..100, computed pass / (pass+warn+fail) * 100


class ProjectComplianceReportList(BaseModel):
    items: list[ProjectComplianceReportItem]
